runtime/parachains/src/router/dmp.rs

```diff
@@ -0,0 +1,212 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::{Trait, Module, Store};
+use frame_support::{StorageMap, weights::Weight, traits::Get};
+use sp_runtime::traits::{BlakeTwo256, Hash as HashT, SaturatedConversion};
+use primitives::v1::{Id as ParaId, DownwardMessage, InboundDownwardMessage, Hash};
+
+/// Routines and getters related to downward message passing.
+impl<T: Trait> Module<T> {
+	pub(crate) fn clean_dmp_after_outgoing(outgoing_para: ParaId) {
+		<Self as Store>::DownwardMessageQueues::remove(&outgoing_para);
+		<Self as Store>::DownwardMessageQueueHeads::remove(&outgoing_para);
+	}
+
+	/// Enqueue a downward message to a specific recipient para.
+	pub fn queue_downward_message(para: ParaId, msg: DownwardMessage) {
+		let inbound = InboundDownwardMessage {
+			msg,
+			sent_at: <frame_system::Module<T>>::block_number(),
+		};
+
+		// obtain the new link in the MQC and update the head.
+		<Self as Store>::DownwardMessageQueueHeads::mutate(para, |head| {
+			let new_head =
+				BlakeTwo256::hash_of(&(*head, inbound.sent_at, T::Hashing::hash_of(&inbound.msg)));
```

Comment:
Is the internal hash (`hash_of(&inbound.msg)`) to be able to traverse hash-chain links without having to provide (potentially large) message content?

Reply:
Yeah, so that the collator can provide the MQC proof to the latest (or perhaps nearest) MQC head beacon.

After all, one DMP message can be up to `{max, critical}_downward_message_size`, and I think it would be a good option to have ability to assign a value that is a big portion of a PoV: all downward messages are under control of the relay-chain anyway and more flexibility won't harm on this.